File: src/ui/containers.ts
import { withAlpha } from '../gfx/color';
import { inset, rect, Rect } from '../gfx/geom';
import type { DisplayList } from '../gfx/painter';
import { theme } from './theme';
import { contains, layoutSubtree, Size, Widget } from './widget';

/** height a child wants at width w: explicit > height-for-width > fallback */
const heightAt = (c: Widget, w: number, fallback: number): number =>
  c.height ?? c.heightForWidth(Math.max(0, w)) ?? fallback;

/**
 * WinForms/WPF-style DockPanel: children consume space from the remaining
 * rect in declaration order according to their `dock` side. a 'fill' child
 * takes whatever is left
 */
export class Dock extends Widget {
  padding = 0;
  gap = 0;

  override layoutChildren(): void {
    this.adopt();
    const rem: Rect = inset(this.bounds, this.padding);
    for (const c of this.children) {
      const sz = this.sizeOf(c);
      let b: Rect;
      switch (c.dock) {
        case 'top': {
          const h = Math.min(heightAt(c, rem.w, sz.h), rem.h);
          b = rect(rem.x, rem.y, rem.w, h);
          rem.y += h + this.gap;
          rem.h = Math.max(0, rem.h - h - this.gap);
          break;
        }
        case 'bottom': {
          const h = Math.min(heightAt(c, rem.w, sz.h), rem.h);
          b = rect(rem.x, rem.y + rem.h - h, rem.w, h);
          rem.h = Math.max(0, rem.h - h - this.gap);
          break;
        }
        case 'left': {
          const w = Math.min(sz.w, rem.w);
          b = rect(rem.x, rem.y, w, rem.h);
          rem.x += w + this.gap;
          rem.w = Math.max(0, rem.w - w - this.gap);
          break;
        }
        case 'right': {
          const w = Math.min(sz.w, rem.w);
          b = rect(rem.x + rem.w - w, rem.y, w, rem.h);
          rem.w = Math.max(0, rem.w - w - this.gap);
          break;
        }
        case 'fill':
          b = rect(rem.x, rem.y, rem.w, rem.h);
          break;
      }
      c.bounds = b;
      layoutSubtree(c);
    }
  }
}

export type StackAlign = 'start' | 'center' | 'end' | 'stretch';

/**
 * NSStackView-style run of children along one axis. Main-axis size per child
 * comes from its `stackSize` (content | fixed | fill-by-weight). the cross
 * axis follows `align`
 */
export class Stack extends Widget {
  padding = 0;
  spacing = 8;
  align: StackAlign = 'start';

  constructor(readonly axis: 'h' | 'v') {
    super();
  }

  override intrinsicSize(): Size {
    this.adopt();
    let main = this.spacing * Math.max(0, this.children.length - 1);
    let cross = 0;
    for (const c of this.children) {
      const sz = this.sizeOf(c);
      const ss = c.stackSize;
      main += ss.kind === 'fixed' ? ss.px : this.axis === 'h' ? sz.w : sz.h;
      cross = Math.max(cross, this.axis === 'h' ? sz.h : sz.w);
    }
    const pad = this.padding * 2;
    return this.axis === 'h' ? { w: main + pad, h: cross + pad } : { w: cross + pad, h: main + pad };
  }

  override layoutChildren(): void {
    this.adopt();
    if (!this.children.length) return;
    const r = inset(this.bounds, this.padding);
    const mainAvail = this.axis === 'h' ? r.w : r.h;
    const crossAvail = this.axis === 'h' ? r.h : r.w;

    let used = this.spacing * (this.children.length - 1);
    let weightSum = 0;
    const mains: number[] = [];
    for (const c of this.children) {
      const ss = c.stackSize;
      if (ss.kind === 'fill') {
        mains.push(0);
        weightSum += ss.weight;
      } else {
        const sz = this.sizeOf(c);
        let m: number;
        if (ss.kind === 'fixed') m = ss.px;
        else if (this.axis === 'h') m = sz.w;
        else {
          // a v-stack knows each child's eventual width, so wrapped labels
          // can restate their height before the main-axis pass
          const cw = this.align === 'stretch' ? crossAvail : Math.min(sz.w, crossAvail);
          m = heightAt(c, cw, sz.h);
        }
        mains.push(m);
        used += m;
      }
    }
    const leftover = Math.max(0, mainAvail - used);

    let pos = this.axis === 'h' ? r.x : r.y;
    this.children.forEach((c, i) => {
      const ss = c.stackSize;
      const main = ss.kind === 'fill' ? (weightSum > 0 ? (leftover * ss.weight) / weightSum : 0) : mains[i]!;
      const sz = this.sizeOf(c);
      // cross-axis children never exceed the stack's extent (labels truncate
      // at the boundary instead of overflowing)
      const cross =
        this.align === 'stretch' ? crossAvail : Math.min(this.axis === 'h' ? sz.h : sz.w, crossAvail);
      let crossPos = this.axis === 'h' ? r.y : r.x;
      if (this.align === 'center') crossPos += (crossAvail - cross) / 2;
      else if (this.align === 'end') crossPos += crossAvail - cross;
      c.bounds = this.axis === 'h' ? rect(pos, crossPos, main, cross) : rect(crossPos, pos, cross, main);
      pos += main + this.spacing;
      layoutSubtree(c);
    });
  }
}

export class HStack extends Stack {
  constructor() {
    super('h');
  }
}

export class VStack extends Stack {
  constructor() {
    super('v');
  }
}

/**
 * Base for anything with a vertical scroll offset: clamping, wheel deltas
 * (routed here by the UI shell), and the draggable overlay thumb. Subclasses 
 * define the scrolled viewport and set contentH during layout.
 */
export abstract class Scroller extends Widget {
  scrollY = 0;
  scrollX = 0;
  protected contentH = 0;
  protected contentW = 0;
  protected dragGrab: number | null = null;
  protected dragGrabX: number | null = null;

  /** the region content scrolls within (a table excludes its header) */
  protected viewport(): Rect {
    return this.bounds;
  }

  protected get maxScroll(): number {
    return Math.max(0, this.contentH - this.viewport().h);
  }

  protected get maxScrollX(): number {
    return Math.max(0, this.contentW - this.viewport().w);
  }

  get scrollable(): boolean {
    return this.maxScroll > 0.5;
  }

  get scrollableX(): boolean {
    return this.maxScrollX > 0.5;
  }

  protected clampScroll(): void {
    this.scrollY = Math.min(Math.max(this.scrollY, 0), this.maxScroll);
    this.scrollX = Math.min(Math.max(this.scrollX, 0), this.maxScrollX);
  }

  scrollBy(dy: number): void {
    const next = Math.min(Math.max(this.scrollY + dy, 0), this.maxScroll);
    if (next !== this.scrollY) {
      this.scrollY = next;
      this.invalidate();
    }
  }

  scrollByX(dx: number): void {
    const next = Math.min(Math.max(this.scrollX + dx, 0), this.maxScrollX);
    if (next !== this.scrollX) {
      this.scrollX = next;
      this.invalidate();
    }
  }

  protected thumbRect(): Rect | null {
    if (!this.scrollable) return null;
    const v = this.viewport();
    const track = v.h - 8;
    const h = Math.max(28, track * (v.h / this.contentH));
    const y = v.y + 4 + (this.scrollY / this.maxScroll) * (track - h);
    return rect(v.x + v.w - 10, y, 6, h);
  }

  protected hThumbRect(): Rect | null {
    if (!this.scrollableX) return null;
    const v = this.viewport();
    const track = v.w - 8;
    const w = Math.max(28, track * (v.w / this.contentW));
    const x = v.x + 4 + (this.scrollX / this.maxScrollX) * (track - w);
    return rect(x, v.y + v.h - 10, w, 6);
  }

  override paintOverlay(dl: DisplayList): void {
    const t = this.thumbRect();
    if (t) dl.rect(t, { color: withAlpha(theme.ink, this.dragGrab != null ? 0.5 : 0.25), radius: 3 });
    const ht = this.hThumbRect();
    if (ht) dl.rect(ht, { color: withAlpha(theme.ink, this.dragGrabX != null ? 0.5 : 0.25), radius: 3 });
  }

  protected thumbDown(x: number, y: number): boolean {
    const t = this.thumbRect();
    if (t && contains(t, x, y)) {
      this.dragGrab = y - t.y;
      this.invalidate();
      return true;
    }
    const ht = this.hThumbRect();
    if (ht && contains(ht, x, y)) {
      this.dragGrabX = x - ht.x;
      this.invalidate();
      return true;
    }
    return false;
  }

  protected thumbDrag(x: number, y: number): boolean {
    if (this.dragGrab != null) {
      const v = this.viewport();
      const track = v.h - 8;
      const h = Math.max(28, track * (v.h / this.contentH));
      const range = track - h;
      if (range > 0) {
        const ratio = (y - this.dragGrab - (v.y + 4)) / range;
        const next = Math.min(Math.max(ratio * this.maxScroll, 0), this.maxScroll);
        if (next !== this.scrollY) {
          this.scrollY = next;
          this.invalidate();
        }
      }
      return true;
    }
    if (this.dragGrabX != null) {
      const v = this.viewport();
      const track = v.w - 8;
      const w = Math.max(28, track * (v.w / this.contentW));
      const range = track - w;
      if (range > 0) {
        const ratio = (x - this.dragGrabX - (v.x + 4)) / range;
        const next = Math.min(Math.max(ratio * this.maxScrollX, 0), this.maxScrollX);
        if (next !== this.scrollX) {
          this.scrollX = next;
          this.invalidate();
        }
      }
      return true;
    }
    return false;
  }

  protected thumbUp(): void {
    if (this.dragGrab != null || this.dragGrabX != null) {
      this.dragGrab = null;
      this.dragGrabX = null;
      this.invalidate();
    }
  }
}

/**
 * Vertically scrollable viewport for ordinary widget children. Children lay
 * out (frame/anchors) against a virtual content rect whose height is their
 * combined extent. The scroll offset just shifts it.
 */
export class ScrollView extends Scroller {
  /** Extra space below the last child. */
  contentInsetBottom = 0;

  constructor() {
    super();
    this.clips = true;
  }

  override layoutChildren(): void {
    this.adopt();
    let extent = 0;
    let extentW = 0;
    for (const c of this.children) {
      const sz = this.sizeOf(c);
      const top = c.frame ? c.frame.y : c.anchors?.top ?? 0;
      const a = c.anchors;
      const wCtx = a && a.left != null && a.right != null ? this.bounds.w - a.left - a.right : sz.w;
      const h = c.frame ? c.frame.h : heightAt(c, wCtx, sz.h);
      extent = Math.max(extent, top + h);
      // width extent from left-pinned/framed children only. right-anchored
      // children track the viewport, not the content.
      const left = c.frame ? c.frame.x : a?.left ?? 0;
      const w = c.frame ? c.frame.w : sz.w;
      if (!a || a.right == null) extentW = Math.max(extentW, left + w);
    }
    this.contentH = extent + this.contentInsetBottom;
    this.contentW = extentW;
    this.clampScroll();
    this.layoutChildrenInto(
      rect(
        this.bounds.x - this.scrollX,
        this.bounds.y - this.scrollY,
        Math.max(this.contentW, this.bounds.w),
        Math.max(this.contentH, this.bounds.h),
      ),
    );
  }

  override hitTest(x: number, y: number): Widget | null {
    if (!contains(this.bounds, x, y)) return null;
    const t = this.thumbRect();
    if (t && contains(t, x, y)) return this; // thumbs paint on top, so they hit first
    const ht = this.hThumbRect();
    if (ht && contains(ht, x, y)) return this;
    for (let i = this.children.length - 1; i >= 0; i--) {
      const hit = this.children[i]!.hitTest(x, y);
      if (hit) return hit;
    }
    return null;
  }

  override onPointerDown(x: number, y: number): void {
    this.thumbDown(x, y);
  }

  override onPointerDrag(x: number, y: number): void {
    this.thumbDrag(x, y);
  }

  override onPointerUp(): void {
    this.thumbUp();
  }
}

File: go/terminal/ui/containers.go
package ui

import (
	"github.com/nullentropy/caution/go/terminal/gfx"
)

// heightAt is the height a child wants at width w: explicit > height-for-
// width > fallback.
func heightAt(c Widget, w, fallback float32) float32 {
	cb := c.Base()
	if cb.Height != nil {
		return *cb.Height
	}
	if h, ok := c.HeightForWidth(max(0, w)); ok {
		return h
	}
	return fallback
}

// Dock is a WinForms/WPF-style DockPanel: children consume space from the
// remaining rect in declaration order according to their `dock` side; a
// 'fill' child takes whatever is left.
type Dock struct {
	Core
	Padding float32
	Gap     float32
}

func NewDock() *Dock { d := &Dock{}; d.self = d; return d }

func (d *Dock) LayoutChildren() {
	d.adopt()
	rem := gfx.Inset(d.Bounds, d.Padding)
	for _, c := range d.Kids {
		cb := c.Base()
		sz := sizeOf(c)
		var b gfx.Rect
		switch cb.Dock {
		case "top":
			h := min(heightAt(c, rem.W, sz.H), rem.H)
			b = gfx.R(rem.X, rem.Y, rem.W, h)
			rem.Y += h + d.Gap
			rem.H = max(0, rem.H-h-d.Gap)
		case "bottom":
			h := min(heightAt(c, rem.W, sz.H), rem.H)
			b = gfx.R(rem.X, rem.Y+rem.H-h, rem.W, h)
			rem.H = max(0, rem.H-h-d.Gap)
		case "left":
			w := min(sz.W, rem.W)
			b = gfx.R(rem.X, rem.Y, w, rem.H)
			rem.X += w + d.Gap
			rem.W = max(0, rem.W-w-d.Gap)
		case "right":
			w := min(sz.W, rem.W)
			b = gfx.R(rem.X+rem.W-w, rem.Y, w, rem.H)
			rem.W = max(0, rem.W-w-d.Gap)
		default: // fill
			b = rem
		}
		cb.Bounds = b
		layoutSubtree(c)
	}
}

type StackAlign = string // "start" (default) | "center" | "end" | "stretch"

// Stack is an NSStackView-style run of children along one axis. Main-axis
// size per child comes from its StackSize (content | fixed | fill-by-weight);
// the cross axis follows Align. No wrapping, no flex algebra.
type Stack struct {
	Core
	Axis    string // "h" | "v"
	Padding float32
	Spacing float32
	Align   StackAlign
}

func NewHStack() *Stack { s := &Stack{Axis: "h", Spacing: 8}; s.self = s; return s }
func NewVStack() *Stack { s := &Stack{Axis: "v", Spacing: 8}; s.self = s; return s }

func (s *Stack) IntrinsicSize() Size {
	s.adopt()
	main := s.Spacing * float32(max(0, len(s.Kids)-1))
	var cross float32
	for _, c := range s.Kids {
		sz := sizeOf(c)
		ss := c.Base().StackSize
		if ss.Kind == "fixed" {
			main += ss.Px
		} else if s.Axis == "h" {
			main += sz.W
		} else {
			main += sz.H
		}
		if s.Axis == "h" {
			cross = max(cross, sz.H)
		} else {
			cross = max(cross, sz.W)
		}
	}
	pad := s.Padding * 2
	if s.Axis == "h" {
		return Size{main + pad, cross + pad}
	}
	return Size{cross + pad, main + pad}
}

func (s *Stack) LayoutChildren() {
	s.adopt()
	if len(s.Kids) == 0 {
		return
	}
	r := gfx.Inset(s.Bounds, s.Padding)
	mainAvail, crossAvail := r.W, r.H
	if s.Axis == "v" {
		mainAvail, crossAvail = r.H, r.W
	}

	used := s.Spacing * float32(len(s.Kids)-1)
	var weightSum float32
	mains := make([]float32, len(s.Kids))
	for i, c := range s.Kids {
		ss := c.Base().StackSize
		if ss.Kind == "fill" {
			weightSum += ss.Weight
			continue
		}
		sz := sizeOf(c)
		m := sz.W
		if s.Axis == "v" {
			// A v-stack knows each child's eventual width, so wrapped labels
			// can restate their height before the main-axis pass.
			cw := min(sz.W, crossAvail)
			if s.Align == "stretch" {
				cw = crossAvail
			}
			m = heightAt(c, cw, sz.H)
		}
		if ss.Kind == "fixed" {
			m = ss.Px
		}
		mains[i] = m
		used += m
	}
	leftover := max(0, mainAvail-used)

	pos := r.X
	if s.Axis == "v" {
		pos = r.Y
	}
	for i, c := range s.Kids {
		ss := c.Base().StackSize
		main := mains[i]
		if ss.Kind == "fill" && weightSum > 0 {
			main = leftover * ss.Weight / weightSum
		}
		sz := sizeOf(c)
		// Cross-axis children never exceed the stack's extent (labels
		// truncate at the boundary instead of overflowing).
		crossNat := sz.H
		if s.Axis == "v" {
			crossNat = sz.W
		}
		cross := min(crossNat, crossAvail)
		if s.Align == "stretch" {
			cross = crossAvail
		}
		crossPos := r.Y
		if s.Axis == "v" {
			crossPos = r.X
		}
		switch s.Align {
		case "center":
			crossPos += (crossAvail - cross) / 2
		case "end":
			crossPos += crossAvail - cross
		}
		if s.Axis == "h" {
			c.Base().Bounds = gfx.R(pos, crossPos, main, cross)
		} else {
			c.Base().Bounds = gfx.R(crossPos, pos, cross, main)
		}
		pos += main + s.Spacing
		layoutSubtree(c)
	}
}

// Scroller is the base for anything with a vertical scroll offset: clamping,
// wheel deltas (routed here by the Ui shell), and the draggable overlay
// thumb. Embedders define the scrolled viewport via viewportFn and set
// ContentH during layout.
type Scroller struct {
	Core
	ScrollY  float32
	ScrollX  float32
	ContentH float32
	ContentW float32

	dragGrab   *float32
	dragGrabX  *float32
	viewportFn func() gfx.Rect
}

// ScrollTarget is what the Ui shell's wheel routing looks for.
type ScrollTarget interface {
	Widget
	ScrollBy(dy float32)
	Scrollable() bool
}

// HScrollTarget is the horizontal half; Scroller implements both.
type HScrollTarget interface {
	ScrollByX(dx float32)
	ScrollableX() bool
}

func (s *Scroller) viewport() gfx.Rect {
	if s.viewportFn != nil {
		return s.viewportFn()
	}
	return s.Bounds
}

func (s *Scroller) maxScroll() float32 {
	return max(0, s.ContentH-s.viewport().H)
}

func (s *Scroller) maxScrollX() float32 {
	return max(0, s.ContentW-s.viewport().W)
}

func (s *Scroller) Scrollable() bool  { return s.maxScroll() > 0.5 }
func (s *Scroller) ScrollableX() bool { return s.maxScrollX() > 0.5 }

func (s *Scroller) clampScroll() {
	s.ScrollY = min(max(s.ScrollY, 0), s.maxScroll())
	s.ScrollX = min(max(s.ScrollX, 0), s.maxScrollX())
}

func (s *Scroller) ScrollBy(dy float32) {
	next := min(max(s.ScrollY+dy, 0), s.maxScroll())
	if next != s.ScrollY {
		s.ScrollY = next
		s.Invalidate()
	}
}

func (s *Scroller) ScrollByX(dx float32) {
	next := min(max(s.ScrollX+dx, 0), s.maxScrollX())
	if next != s.ScrollX {
		s.ScrollX = next
		s.Invalidate()
	}
}

func (s *Scroller) thumbRect() (gfx.Rect, bool) {
	if !s.Scrollable() {
		return gfx.Rect{}, false
	}
	v := s.viewport()
	track := v.H - 8
	h := max(28, track*(v.H/s.ContentH))
	y := v.Y + 4 + (s.ScrollY/s.maxScroll())*(track-h)
	return gfx.R(v.X+v.W-10, y, 6, h), true
}

func (s *Scroller) hThumbRect() (gfx.Rect, bool) {
	if !s.ScrollableX() {
		return gfx.Rect{}, false
	}
	v := s.viewport()
	track := v.W - 8
	w := max(28, track*(v.W/s.ContentW))
	x := v.X + 4 + (s.ScrollX/s.maxScrollX())*(track-w)
	return gfx.R(x, v.Y+v.H-10, w, 6), true
}

func (s *Scroller) PaintOverlay(dl *gfx.DisplayList) {
	if t, ok := s.thumbRect(); ok {
		alpha := float32(0.25)
		if s.dragGrab != nil {
			alpha = 0.5
		}
		dl.Fill(t, gfx.WithAlpha(*Theme["ink"], alpha), gfx.CornerRadius(3))
	}
	if t, ok := s.hThumbRect(); ok {
		alpha := float32(0.25)
		if s.dragGrabX != nil {
			alpha = 0.5
		}
		dl.Fill(t, gfx.WithAlpha(*Theme["ink"], alpha), gfx.CornerRadius(3))
	}
}

func (s *Scroller) thumbDown(x, y float32) bool {
	if t, ok := s.thumbRect(); ok && Contains(t, x, y) {
		grab := y - t.Y
		s.dragGrab = &grab
		s.Invalidate()
		return true
	}
	if t, ok := s.hThumbRect(); ok && Contains(t, x, y) {
		grab := x - t.X
		s.dragGrabX = &grab
		s.Invalidate()
		return true
	}
	return false
}

func (s *Scroller) thumbDrag(x, y float32) bool {
	if s.dragGrab != nil {
		v := s.viewport()
		track := v.H - 8
		h := max(28, track*(v.H/s.ContentH))
		rng := track - h
		if rng > 0 {
			ratio := (y - *s.dragGrab - (v.Y + 4)) / rng
			next := min(max(ratio*s.maxScroll(), 0), s.maxScroll())
			if next != s.ScrollY {
				s.ScrollY = next
				s.Invalidate()
			}
		}
		return true
	}
	if s.dragGrabX != nil {
		v := s.viewport()
		track := v.W - 8
		w := max(28, track*(v.W/s.ContentW))
		rng := track - w
		if rng > 0 {
			ratio := (x - *s.dragGrabX - (v.X + 4)) / rng
			next := min(max(ratio*s.maxScrollX(), 0), s.maxScrollX())
			if next != s.ScrollX {
				s.ScrollX = next
				s.Invalidate()
			}
		}
		return true
	}
	return false
}

func (s *Scroller) thumbUp() {
	if s.dragGrab != nil || s.dragGrabX != nil {
		s.dragGrab = nil
		s.dragGrabX = nil
		s.Invalidate()
	}
}

// ScrollView is a vertically scrollable viewport for ordinary widget
// children. Children lay out (frame/anchors) against a virtual content rect
// whose height is their combined extent; the scroll offset just shifts it.
type ScrollView struct {
	Scroller
	// ContentInsetBottom is extra space below the last child.
	ContentInsetBottom float32
}

func NewScrollView() *ScrollView {
	sv := &ScrollView{}
	sv.self = sv
	sv.Clips = true
	return sv
}

func (sv *ScrollView) LayoutChildren() {
	sv.adopt()
	var extent, extentW float32
	for _, c := range sv.Kids {
		cb := c.Base()
		sz := sizeOf(c)
		var top, h float32
		if cb.Frame != nil {
			top, h = cb.Frame.Y, cb.Frame.H
		} else {
			wCtx := sz.W
			if a := cb.Anchors; a != nil {
				if a.Top != nil {
					top = *a.Top
				}
				if a.Left != nil && a.Right != nil {
					wCtx = sv.Bounds.W - *a.Left - *a.Right
				}
			}
			h = heightAt(c, wCtx, sz.H)
		}
		extent = max(extent, top+h)
		// Width extent from left-pinned/framed children only - right-anchored
		// children track the viewport, not the content.
		var left, w float32
		if cb.Frame != nil {
			left, w = cb.Frame.X, cb.Frame.W
		} else {
			w = sz.W
			if a := cb.Anchors; a != nil {
				if a.Right != nil {
					w = 0 // tracks the viewport
				}
				if a.Left != nil {
					left = *a.Left
				}
			}
		}
		extentW = max(extentW, left+w)
	}
	sv.ContentH = extent + sv.ContentInsetBottom
	sv.ContentW = extentW
	sv.clampScroll()
	sv.LayoutInto(gfx.R(sv.Bounds.X-sv.ScrollX, sv.Bounds.Y-sv.ScrollY,
		max(sv.ContentW, sv.Bounds.W), max(sv.ContentH, sv.Bounds.H)))
}

func (sv *ScrollView) HitTest(x, y float32) Widget {
	if !Contains(sv.Bounds, x, y) {
		return nil
	}
	if t, ok := sv.thumbRect(); ok && Contains(t, x, y) {
		return sv // thumbs paint on top, so they hit first
	}
	if t, ok := sv.hThumbRect(); ok && Contains(t, x, y) {
		return sv
	}
	for i := len(sv.Kids) - 1; i >= 0; i-- {
		if hit := sv.Kids[i].HitTest(x, y); hit != nil {
			return hit
		}
	}
	return nil
}

func (sv *ScrollView) OnPointerDown(x, y float32, _ int) { sv.thumbDown(x, y) }
func (sv *ScrollView) OnPointerDrag(x, y float32)        { sv.thumbDrag(x, y) }
func (sv *ScrollView) OnPointerUp(_, _ float32)          { sv.thumbUp() }
